Public/Src/Tools/SymbolDaemon/VsoSymbolClient.cs

```diff
@@ -0,0 +1,208 @@
+        private readonly ISymbolServiceClient m_symbolClient;
+        private readonly CancellationTokenSource m_cancellationSource;
+        private CancellationToken Token => m_cancellationSource.Token;
+        private string m_requestId;
+
+        private VssCredentials GetCredentials() =>
+            new VsoCredentialHelper(m => m_logger.Verbose(m))
+                .GetCredentials(m_config.Service, true, null);
+
+        private ArtifactHttpClientFactory GetFactory() =>
+            new ArtifactHttpClientFactory(
+                credentials: GetCredentials(),
+                httpSendTimeout: m_config.HttpSendTimeout,
+                tracer: Tracer,
+                verifyConnectionCancellationToken: Token);
+
+        private Uri ServiceEndpoint => m_config.Service;
+
+        private string RequestName => m_config.Name;
+
+        private string RequestId
+        {
+            get
+            {
+                Contract.Requires(!string.IsNullOrEmpty(m_requestId));
+                return m_requestId;
+            }
+        }
+
+        /// <nodoc />
+        public VsoSymbolClient(ILogger logger, SymbolConfig config)
+        {
+            m_logger = logger;
+            m_config = config;
+            m_cancellationSource = new CancellationTokenSource();
+
+            m_logger.Info(I($"[{nameof(VsoSymbolClient)}] Using drop config: {JsonConvert.SerializeObject(m_config)}"));
+
+            m_symbolClient = CreateSymbolServiceClient();
+        }
+
+        private ISymbolServiceClient CreateSymbolServiceClient()
+        {
+            var client = new SymbolServiceClient(
+                ServiceEndpoint,
+                GetFactory(),
+                Tracer,
+                new SymbolServiceClientTelemetry(Tracer, ServiceEndpoint, enable: m_config.EnableTelemetry));
+
+            return client;
+        }
+
+        /// <summary>
+        /// Queries the endpoint for the RequestId. 
+        /// This method should be called only after the request has been created, otherwise, it will throw an exception.
+        /// </summary>
+        /// <remarks>
+        /// On workers, m_requestId won't be initialized, so we need to query the server for the right value.
+        /// </remarks>
+        private async Task EnsureRequestIdInitalizedAsync()
+        {
+            if (string.IsNullOrEmpty(m_requestId))
+            {
+                var result = await m_symbolClient.GetRequestByNameAsync(RequestName, Token);
+                m_requestId = result.Id;
+            }
+        }
+
+        /// <summary>
+        /// Creates a symbol request.
+        /// </summary>
+        public async Task<Request> CreateAsync(CancellationToken token)
+        {
+            var result = await m_symbolClient.CreateRequestAsync(RequestName, token);
+
+            m_requestId = result.Id;
+
+            return result;
+        }
+
+        /// <inheritdoc />
+        public Task<Request> CreateAsync() => CreateAsync(Token);

```

Comment:
Who may issue the cancellation?

Reply:
Nobody. It's a required argument from the API point of view though, so we create and pass one.

---
In reply to: [313642210](https://github.com/microsoft/BuildXL/pull/727#discussion_r313642210) [](ancestors = 313642210)